### PR TITLE
Feat/add song

### PR DESCRIPTION
**Pull request by KlaraHajdu**
Tasks: 

* add AddSong component to PlaylistPage
* add AddSong async action and firestoreApi function
* add subscribe and unsubscribe from songs collection on current playlist
* add unsubscribe from current playlist

### CODE REVIEW

File: src/components/AddSong/AddSong.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            setUrl("");
+        } else {
+            setUrl(url);
+            setHelperText(null);
+        }
+    };
+
+    const submitSongUrl = (event: FormEvent<HTMLButtonElement>) => {
+        event.preventDefault();
+        console.log("will disp");
```
Comment (bbucsek) [inline src/components/AddSong/AddSong.tsx]: delete clg pls

File: src/components/AddSong/AddSong.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        if (helperText ) {
+            return;
+        }
+
+        if (url.length === 0) {
```
Comment (bbucsek) [inline src/components/AddSong/AddSong.tsx:28]: if the url lenght is 0, then the helpertext is not null right? so line is redundant

File: src/components/AddSong/AddSong.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            return
+        }
+        dispatch(playlistsAsyncActions.verifyUrl(url));
+        setUrl("");
+    };
+
+    return (
+        <Wrapper>
+            <Container>
```
Comment (bbucsek) [inline src/components/AddSong/AddSong.tsx]: wrapper <-> container

File: src/components/AddSong/styles.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import styled from 'styled-components'
+
+export const Wrapper = styled.div`
+    margin: auto;
```
Comment (bbucsek) [inline src/components/AddSong/styles.ts]: what about delete all the margin:auto and replace with align-items?

File: src/service/firestoreApi.ts [hunk 5]
```diff
@@ -9,6 +10,7 @@ const createPlaylist = async (owner: string, playlistName: string) => {
 const subscribeToPlaylist = async (id: string, observer: (playlist: any) => void) => {
     const callback = (snapshot: any) => {
         const playlistData = snapshot.data()
+        console.log(playlistData)
```
Comment (bbucsek) [inline src/service/firestoreApi.ts]: delete pls

File: src/service/firestoreApi.ts [hunk 6]
```diff
@@ -19,22 +21,41 @@ const subscribeToPlaylist = async (id: string, observer: (playlist: any) => void
...
+            });
+            return data
 
-    })
+        })
+}
+
+const subscribeToSongsCollection = async (id: string, observer: (playlist: any) => void) => {
+    const callback = (snapshot: any) => {
+        console.log(snapshot.docs)
```
Comment (bbucsek) [inline src/service/firestoreApi.ts]: clg

File: src/service/firestoreApi.ts [hunk 7]
```diff
@@ -19,22 +21,41 @@ const subscribeToPlaylist = async (id: string, observer: (playlist: any) => void
...
+        const songList = snapshot.docs.map((doc: any) => {
+            const song = { ...doc.data(), id: doc.id }
+            return song
+        })
+        console.log(songList)
+        observer(songList)
+    }
+    await database.collection('playlists').doc(id).collection('songs').onSnapshot(callback)
+}
```
Comment (bbucsek) [inline src/service/firestoreApi.ts:63]: can you seperate the chained funcitons with a new line?
.foo()
.bar()
.baz()

File: src/store/slices/playlists/slice.ts [hunk 8]
```diff
@@ -30,12 +33,77 @@ string,
...
+    { state: RootState }
+    >
+    ('playlists/verifyUrl',
+        async (payload: string, thunkApi) => {
+            const state = thunkApi.getState()
+
+            const url = payload
+            const youtubeId = getYoutubeId(url)
+            if (!youtubeId) {
+                return "not_valid_youtube_url"
```
Comment (bbucsek) [inline src/store/slices/playlists/slice.ts]: not thunkapi.rejectwithvalue?

File: src/store/slices/playlists/slice.ts [hunk 9]
```diff
@@ -30,12 +33,77 @@ string,
...
+            }
+            const videoDetails = await getVideoDetails(youtubeId)
+
+            const videoDurationIsOk = checkIfVideoDurationIsOk(videoDetails.duration)
+
+            if (!videoDurationIsOk) {
+                return "video_too_long"
```
Comment (bbucsek) [inline src/store/slices/playlists/slice.ts]: not thunkapi.rejectwithvalue?

File: src/store/slices/playlists/slice.ts [hunk 10]
```diff
@@ -30,12 +33,77 @@ string,
...
+    {youtubeId: string, title: string},
+    { state: RootState }
+    >
+    ('playlists/addSong',
+        async (payload, thunkApi) => {
+            const state = thunkApi.getState()
+            const { authentication } = state
+            const { currentUser } = authentication
+            if (!currentUser) {
+                return "no_currentUser"
```
Comment (bbucsek) [inline src/store/slices/playlists/slice.ts]: not thunkapi.rejectwithvalue?

File: src/store/slices/playlists/slice.ts [hunk 11]
```diff
@@ -30,12 +33,77 @@ string,
...
+            }
+            const userId = currentUser.id
+            
+            const { playlists } = state
+            const { currentPlaylist } = playlists
+            if (!currentPlaylist) {
+                return "no_currentPlaylist"
```
Comment (bbucsek) [inline src/store/slices/playlists/slice.ts]: not thunkapi.rejectwithvalue?

File: src/store/slices/playlists/slice.ts [hunk 12]
```diff
@@ -106,12 +177,36 @@ const subscribeToPlaylist = createAsyncThunk<
...
+    string,
+    string,
+    { state: RootState } >(
+    'playlists/subscribeToSongsCollection',
+        async (payload, thunkApi) => {
+        const id = payload;
+        
+        try {
+            const observer = (songs: Song[]) => {
+            thunkApi.dispatch(slice.actions.SET_SONGS(songs))
```
Comment (bbucsek) [inline src/store/slices/playlists/slice.ts]: indentation pls

File: src/components/AddSong/AddSong.test.tsx [hunk 13]
```diff
@@ -18,4 +18,76 @@ describe("AddPlaylist component", () => {
...
+
+        const input = getByTestId("url-input");
+        await userEvent.type(input, "www.");
+        await userEvent.clear(input);
+        const button = getByTestId("submit-button");
+        await userEvent.click(button);
+
+        expect(mockedDispatch).not.toBeCalled();
+    });
+    it("does not dispatch action when helper text is null, but inpput is empty and button is clicked", async () => {
```
Comment (bbucsek) [inline src/components/AddSong/AddSong.test.tsx]: typo

File: src/store/slices/playlists/slice.test.ts [hunk 14]
```diff
@@ -125,6 +180,29 @@ describe('CreatePlaylist slice async action', () => {
...
+    expect(actions[1].type).toEqual('playlists/verifyUrl/rejected')
+    expect(actions[1].payload).toEqual("no_youtube_url")
+  })
+  fit('returns error action if song is too long', async () => {
+    mockedGetVideoDetails.mockReturnValueOnce({ title: "Title", duration: "PT15M", youtubeId: "fake_youtubeId" })
+    await store.dispatch(playlistsAsyncActions.verifyUrl("https://www.youtube.com/watch?v=rVgkIGzGzaU&ab_channel=SevenBeatsMusic"))
+
+
+    const actions = store.getActions()
+    console.log(actions)
```
Comment (bbucsek) [inline src/store/slices/playlists/slice.test.ts]: clg
also please new line before evaluation

**Review by bbucsek — CHANGES_REQUESTED** [review]

File: src/components/AddSong/AddSong.tsx (hunk 1 above)
Comment (KlaraHajdu) [reply src/components/AddSong/AddSong.tsx]: already deleted :)

File: src/components/AddSong/AddSong.tsx (hunk 2 above)
Comment (KlaraHajdu) [reply src/components/AddSong/AddSong.tsx:28]: When the component mounts, the length is still 0, but there is still no helper text. Only if the user writes sg, and then deletes the input.

File: src/components/AddSong/AddSong.tsx (hunk 3 above)
Comment (KlaraHajdu) [reply src/components/AddSong/AddSong.tsx]: Already done. :)

File: src/components/AddSong/styles.ts (hunk 4 above)
Comment (KlaraHajdu) [reply src/components/AddSong/styles.ts]: fixed.

File: src/service/firestoreApi.ts (hunk 5 above)
Comment (KlaraHajdu) [reply src/service/firestoreApi.ts]: already done :)

File: src/service/firestoreApi.ts (hunk 6 above)
Comment (KlaraHajdu) [reply src/service/firestoreApi.ts]: already done :)

File: src/service/firestoreApi.ts (hunk 7 above)
Comment (KlaraHajdu) [reply src/service/firestoreApi.ts:63]: fixed

File: src/store/slices/playlists/slice.ts (hunk 8 above)
Comment (KlaraHajdu) [reply src/store/slices/playlists/slice.ts]: already done :)

File: src/store/slices/playlists/slice.ts (hunk 9 above)
Comment (KlaraHajdu) [reply src/store/slices/playlists/slice.ts]: already done :)

File: src/store/slices/playlists/slice.ts (hunk 10 above)
Comment (KlaraHajdu) [reply src/store/slices/playlists/slice.ts]: already done :)

File: src/components/AddSong/AddSong.tsx (hunk 2 above)
Comment (bbucsek) [reply src/components/AddSong/AddSong.tsx:28]: maybe set the helpertext to "cannot be null" right away?

File: src/store/slices/playlists/slice.ts (hunk 11 above)
Comment (KlaraHajdu) [reply src/store/slices/playlists/slice.ts]: already done :)

File: src/store/slices/playlists/slice.ts (hunk 12 above)
Comment (KlaraHajdu) [reply src/store/slices/playlists/slice.ts]: fixed

File: src/components/AddSong/AddSong.test.tsx (hunk 13 above)
Comment (KlaraHajdu) [reply src/components/AddSong/AddSong.test.tsx]: fixed

File: src/store/slices/playlists/slice.test.ts (hunk 14 above)
Comment (KlaraHajdu) [reply src/store/slices/playlists/slice.test.ts]: already done :)

**Review by bbucsek — APPROVED** [review]

**Pull request merged**